lib/vintage_net_mobile/modem/automatic/discovery.ex

```diff
@@ -0,0 +1,130 @@
+
+  @quectel_vender_id 0x2C7C
+  @telit_vendor_id 0x1BC7
+
+  @modems %{
+    {@telit_vendor_id, 0x1201} => VintageNetMobile.Modem.TelitLE910,
+    {@quectel_vender_id, 0x0125} => VintageNetMobile.Modem.QuectelEC25,
+    # map the EC21 to the EC25 as they work the same
+    {@quectel_vender_id, 0x0121} => VintageNetMobile.Modem.QuectelEC25,
+    {@quectel_vender_id, 0x0296} => VintageNetMobile.Modem.QuectelBG96
+  }
+
+  @typedoc """
+  Init args
+
+  - `:raw_config` - The RawConfig that contains configuration information to
+    pass the discovered modem
+  """
+  @type init_arg() :: {:raw_config, VintageNet.Interface.RawConfig.t()}
+
+  @doc """
+  Start the discovery server
+  """
+  @spec start_link([init_arg()]) :: GenServer.on_start()
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, name: __MODULE__)
+  end
+
+  @impl GenServer
+  def init(args) do
+    {:ok, %{modem: nil, raw_config: args[:raw_config], backoff: 500}, {:continue, :detect_modem}}
+  end
+
+  @impl GenServer
+  def handle_continue(:detect_modem, state) do
+    {:noreply, do_detect_modem(state)}
+  end
+
+  @impl GenServer
+  def handle_info(:detect_modem, state) do
+    {:noreply, do_detect_modem(state)}
+  end
+
+  def handle_info(:configure_vintage_net, state) do
+    {:noreply, configure_vintage_net(state)}
+  end
+
+  defp do_detect_modem(state) do
+    case detect_modem() do
+      nil ->
+        new_backoff = get_next_backoff(state)
+
+        Logger.warn(
+          "[VintageNetMobile] unable to detect modem trying again in #{inspect(new_backoff)}"
+        )
+
+        Process.send_after(self(), :detect_modem, new_backoff)
+
+        %{state | backoff: new_backoff}
+
+      modem ->
+        state = %{state | modem: modem}
+        configure_vintage_net(state)
+    end
+  end
+
+  defp get_next_backoff(%{backoff: backoff}) when backoff >= @max_backoff do
+    backoff
+  end
+
+  defp get_next_backoff(state), do: state.backoff + 500
+
+  defp configure_vintage_net(state) do
+    config = state.raw_config.source_config.vintage_net_mobile |> Map.put(:modem, state.modem)
+
+    case VintageNet.configure(
+           state.raw_config.ifname,
+           %{type: VintageNetMobile, vintage_net_mobile: config},
+           persist: false
+         ) do
+      :ok ->
+        state
+
+      {:error, reason} ->
+        Logger.warn(
+          "[VintageNetMobile] unable to configure VintageNet because #{inspect(reason)}"
+        )
+
+        Process.send_after(self(), :configure_vintage_net, 1_000)
+        state
+    end
+  end
+
+  defp detect_modem() do
+    UART.enumerate()
+    |> Enum.filter(fn {_tty, uart} -> !Enum.empty?(uart) end)
+    |> detect_modem()
+  end
+
+  defp detect_modem([]), do: nil
+
+  defp detect_modem([{_tty, device} | devices]) do
+    case Map.get(@modems, {device.vendor_id, device.product_id}) do
+      nil ->
+        detect_modem(devices)
+
+      modem ->
+        modem
+    end
+  end
```

Comment:
```suggestion
    |> Enum.find_value(&known_modem/1)
  end

  defp known_modem({_tty, %{vendor_id: vid, product_id: pid}}) do
    Map.get(@modems, {vid, pid})
  end
  
  defp known_modem(_), do: nil
```

This handles the case where the UART returned is not empty, but doesn't have a vid:pid. This exists, but I'm not sure we encounter this much yet.